crates/engine/invalid-block-hooks/src/witness.rs

```diff
@@ -1,13 +1,224 @@
-use reth_primitives::{Receipt, SealedBlockWithSenders, SealedHeader, B256};
-use reth_provider::BlockExecutionOutput;
-use reth_trie::updates::TrieUpdates;
+use std::{collections::HashMap, fmt::Debug, fs::File, io::Write, path::PathBuf};
+
+use alloy_rpc_types_debug::ExecutionWitness;
+use eyre::OptionExt;
+use pretty_assertions::Comparison;
+use reth_chainspec::ChainSpec;
+use reth_engine_primitives::InvalidBlockHook;
+use reth_evm::{
+    system_calls::{apply_beacon_root_contract_call, apply_blockhashes_contract_call},
+    ConfigureEvm,
+};
+use reth_primitives::{keccak256, Receipt, SealedBlockWithSenders, SealedHeader, B256, U256};
+use reth_provider::{BlockExecutionOutput, ChainSpecProvider, StateProviderFactory};
+use reth_revm::{
+    database::StateProviderDatabase,
+    db::states::bundle_state::BundleRetention,
+    primitives::{BlockEnv, CfgEnvWithHandlerCfg, EnvWithHandlerCfg},
+    DatabaseCommit, StateBuilder,
+};
+use reth_tracing::tracing::warn;
+use reth_trie::{updates::TrieUpdates, HashedPostState, HashedStorage};
 
 /// Generates a witness for the given block and saves it to a file.
-pub fn witness(
-    _block: &SealedBlockWithSenders,
-    _header: &SealedHeader,
-    _output: &BlockExecutionOutput<Receipt>,
-    _trie_updates: Option<(&TrieUpdates, B256)>,
-) {
-    unimplemented!("witness generation is not supported")
+#[derive(Debug)]
+pub struct Witness<P, EvmConfig> {
+    /// The directory to write the witness to. Additionally, diff files will be written to this
+    /// directory in case of failed sanity checks.
+    output_directory: PathBuf,
+    /// The provider to read the historical state and do the EVM execution.
+    provider: P,
+    /// The EVM configuration to use for the execution.
+    evm_config: EvmConfig,
+}
+
+impl<P, EvmConfig> Witness<P, EvmConfig> {
+    /// Creates a new witness hook.
+    pub const fn new(output_directory: PathBuf, provider: P, evm_config: EvmConfig) -> Self {
+        Self { output_directory, provider, evm_config }
+    }
+}
+
+impl<P, EvmConfig> Witness<P, EvmConfig> {
+    /// Compares two values and saves the diff if they are different.
+    ///
+    /// Returns the path of the saved diff file if the values are different.
+    fn compare_and_save_diff<T: PartialEq + Debug>(
+        &self,
+        filename: String,
+        original: &T,
+        new: &T,
+    ) -> Option<PathBuf> {
+        if original == new {
+            return None
+        }
```

Comment:
this fn should just be responsible for writing, will be much cleaner if you compare the outputs where u call this fn